packages/contracts/src/contracts/current/forwarder/MixinBuyExactAssets.sol

```diff
@@ -0,0 +1,137 @@
+            require(requestedTokensResult.makerAssetFilledAmount >= assetAmount, UNACCEPTABLE_THRESHOLD);
+        } else {
+            Exchange.FillResults memory calculatedMarketBuyResults = calculateMarketBuyFillResults(orders, assetAmount);
+            if (calculatedMarketBuyResults.takerFeePaid > 0) {
+                // Fees are required for these orders. Buy enough ZRX to cover the future market buy
+                Exchange.FillResults memory feeTokensResult = buyFeeTokensInternal(
+                    feeOrders, feeSignatures, calculatedMarketBuyResults.takerFeePaid);
+                totalFillResults.takerAssetFilledAmount = feeTokensResult.takerAssetFilledAmount;
+                totalFillResults.takerFeePaid = feeTokensResult.takerFeePaid;
+            }
+            // Make our market buy of the requested tokens with the remaining balance
+            requestedTokensResult = EXCHANGE.marketBuyOrders(orders, assetAmount, signatures);
+            require(requestedTokensResult.makerAssetFilledAmount == assetAmount, UNACCEPTABLE_THRESHOLD);
+        }
+        addFillResults(totalFillResults, requestedTokensResult);
+        // Transfer all tokens to msg.sender
+        transferToken(makerTokenAddress, msg.sender, totalFillResults.makerAssetFilledAmount);
+        return totalFillResults;
+    }
+
+    function buyExactERC721TokensInternal(
+        Order[] memory orders,
+        bytes[] memory signatures,
+        Order[] memory feeOrders,
+        bytes[] memory feeSignatures,
+        uint256 assetAmount)
+        private
+        returns (Exchange.FillResults memory totalFillResults)
+    {
+        require(assetAmount == orders.length, ASSET_AMOUNT_MATCH_ORDER_SIZE);
+        uint256 totalFeeAmount;
+        for (uint256 i = 0; i < orders.length; i++) {
+            totalFeeAmount = safeAdd(totalFeeAmount, orders[i].takerFee);
+        }
+        if (totalFeeAmount > 0) {
+            // Fees are required for these orders. Buy enough ZRX to cover the future fill
+            Exchange.FillResults memory feeTokensResult = buyFeeTokensInternal(feeOrders, feeSignatures, totalFeeAmount);
+            totalFillResults.takerFeePaid = feeTokensResult.takerFeePaid;
+            totalFillResults.takerAssetFilledAmount = feeTokensResult.takerAssetFilledAmount;
+        }
+        address makerTokenAddress = readAddress(orders[0].makerAssetData, 0);
+        for (uint256 n = 0; n < orders.length; n++) {
+            // Fail if it wasn't fully filled otherwise we will keep WETH
+            Exchange.FillResults memory fillOrderResults = EXCHANGE.fillOrKillOrder(
+                orders[n],
+                orders[n].takerAssetAmount,
+                signatures[n]
+            );
+            addFillResults(totalFillResults, fillOrderResults);
+            uint256 tokenId = readUint256(orders[n].makerAssetData, 20);
+            transferNFTToken(makerTokenAddress, msg.sender, tokenId);
```

Comment:
Can we inherit from ERC721Proxy and call transferFromInternal?

I realize this is slightly more efficient since we don't have to decode the `makerTokenAddress`, but it would reduce the complexity. Esp once we merge the PR with `receiverData`/`safeTransferFrom`.